<center>
    <img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>


## *Extracting and Visualizing Stock Data*
## *Description*

Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


$ \ $

----



## *Table of Contents*

* Define a Function that Makes a Graph

* Question 1: Use `yfinance` to Extract Stock Data

* Question 2: Use `Webscraping` to Extract Tesla Revenue Data

* Question 3: Use `yfinance` to Extract Stock Data

* Question 4: Use `Webscraping` to Extract GME Revenue Data

* Question 5: Plot Tesla Stock Graph

* Question 6: Plot GameStop Stock Graph

$ \ $

-----


The following are the packages that we will install and use.


In [1]:
#(1) yfinance es una popular biblioteca de código abierto desarrollada por Ran Aroussi como un medio para acceder a los datos financieros
!pip install yfinance==0.1.67

#(2) pandas es una libreria de Python usada para el análisis de datos en tablas.
#!pip install pandas==1.3.3

#(3) requests es una librería Python que facilita enormemente el trabajo con peticiones HTTP.
#!pip install requests==2.26.0

#(4) Beautiful Soup es una librería Python que permite extraer información de contenido en formato HTML o XML. Para usarla, es necesario especificar un parser, que es responsable de transformar un documento HTML o XML en un árbol complejo de objetos Python.
!mamba install bs4==4.10.0 -y

#(5) plotly.plotly es una sublibrería que tiene la función de hacer de enlace entre Plotly y tu máquina local
#!pip install plotly==5.3.1


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
/bin/bash: mamba: command not found


-----

$ \ $

We import some packages from the previously installed libraries.


$ \ $

In [2]:
#(1) importamos la libreria yfinance y reducimos su notación.
import yfinance as yf

#(2) importamos la libreria pandas y reducimos su notación.
import pandas as pd

#(3) importamos la libraria requests.
import requests

#(4) importamos la libreria BeautifulSoup.
from bs4 import BeautifulSoup 

#(5) Diagrama de dispersión en Plotly usando la clase graph_objects. 
import plotly.graph_objects as go

#(6) un poco extraño, pero parece que sirve para obtener graficas de otras graficas
from plotly.subplots import make_subplots


-----

## *Define Graphing Function.*

In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.



$ \ $


-------

In [3]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021--06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()
    

------

$ \ $

## *Question 1: Use `yfinance` to Extract Stock Data.*

Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


$ \ $

In [4]:
Tesla = yf.Ticker("TSLA") 

-----

$ \ $

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.

$ \ $


In [5]:
tesla_data = Tesla.history(period="max") 

--------

$ \ $

## *Reset the index.*

Using the `reset_index(inplace=True)` function on the tesla_data DataFrame to display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


$ \ $



In [28]:
tesla_data.reset_index(inplace=True)
tesla_data.head()

index       Date      Open      High       Low     Close     Volume  \
0      0 2010-06-29  1.266667  1.666667  1.169333  1.592667  281494500   
1      1 2010-06-30  1.719333  2.028000  1.553333  1.588667  257806500   
2      2 2010-07-01  1.666667  1.728000  1.351333  1.464000  123282000   
3      3 2010-07-02  1.533333  1.540000  1.247333  1.280000   77097000   
4      4 2010-07-06  1.333333  1.333333  1.055333  1.074000  103003500   

   Dividends  Stock Splits  
0          0           0.0  
1          0           0.0  
2          0           0.0  
3          0           0.0  
4          0           0.0

$ \ $

-------

## *Question 2: Use Webscraping to Extract Tesla Revenue Data.*

$ \ $

Use the `requests` library to download the webpage

$ \ $

[https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue](https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork23455606-2022-01-01). 
 
 
$ \ $

 Save the text of the response as a variable named `html_data`.



$ \ $

In [7]:
url = "https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue"

html_data = requests.get(url).text


$ \ $

-------

Parse the html data using `beautiful_soup`.

$ \ $


In [8]:
soup = BeautifulSoup(html_data, 'html5lib')

------------------------

$ \ $

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Quarterly Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.




$ \ $


In [9]:
tesla_revenue= pd.read_html(url)[1]
tesla_revenue.head()

Tesla Quarterly Revenue(Millions of US $)  \
0                                2022-06-30   
1                                2022-03-31   
2                                2021-12-31   
3                                2021-09-30   
4                                2021-06-30   

  Tesla Quarterly Revenue(Millions of US $).1  
0                                     $16,934  
1                                     $18,756  
2                                     $17,719  
3                                     $13,757  
4                                     $11,958

<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


------------------

$ \ $

Execute the following line to remove the comma and dollar sign from the `Revenue` column.


$ \ $



In [10]:
tesla_revenue=tesla_revenue.rename(columns={"Tesla Quarterly Revenue(Millions of US $)":"Date","Tesla Quarterly Revenue(Millions of US $).1":"Revenue"})
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")
tesla_revenue.head()


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  


Date Revenue
0  2022-06-30   16934
1  2022-03-31   18756
2  2021-12-31   17719
3  2021-09-30   13757
4  2021-06-30   11958

----

$ \ $ 

Execute the following lines to remove an null or empty strings in the Revenue column.


$ \ $


In [11]:
tesla_revenue.dropna(inplace=True)
tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

$ \ $

-----

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


$ \ $


In [12]:
tesla_revenue.reset_index(inplace=True)
tesla_revenue.tail()

index        Date Revenue
47     47  2010-09-30      31
48     48  2010-06-30      28
49     49  2010-03-31      21
50     51  2009-09-30      46
51     52  2009-06-30      27

$ \ $

------

## *Question 3: Use yfinance to Extract Stock Data*

$ \ $

Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.



$ \ $

In [13]:
GME = yf.Ticker("GME")  

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [26]:
gme_data = GME.history(period="max")

In [27]:
gme_data.reset_index(inplace=True)
gme_data.head()

Date      Open      High       Low     Close    Volume  Dividends  \
0 2002-02-13  1.620129  1.693350  1.603296  1.691667  76216000        0.0   
1 2002-02-14  1.712707  1.716073  1.670626  1.683250  11021600        0.0   
2 2002-02-15  1.683250  1.687458  1.658002  1.674834   8389600        0.0   
3 2002-02-19  1.666418  1.666418  1.578047  1.607504   7410400        0.0   
4 2002-02-20  1.615920  1.662210  1.603296  1.662210   6892800        0.0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0

$ \ $

-------

## *Reset the index*

Using the `reset_index(inplace=True)` function on the `gme_data` DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


$ \ $


In [15]:
gme_data.reset_index(inplace=True)
gme_data.head()

Date      Open      High       Low     Close    Volume  Dividends  \
0 2002-02-13  1.620128  1.693350  1.603296  1.691666  76216000        0.0   
1 2002-02-14  1.712707  1.716074  1.670626  1.683250  11021600        0.0   
2 2002-02-15  1.683250  1.687458  1.658002  1.674834   8389600        0.0   
3 2002-02-19  1.666418  1.666418  1.578047  1.607504   7410400        0.0   
4 2002-02-20  1.615921  1.662210  1.603296  1.662210   6892800        0.0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0

$ \ $

--------


## *Question 4: Use Webscraping to Extract GME Revenue Data.*

$ \ $

Use the `requests` library to download the webpage 

$ \ $

<https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html>.

$ \ $

Save the text of the response as a variable named `html_data`.


$ \ $

In [16]:
direction = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"
html_data = requests.get(direction).text


$ \ $

------

Parse the html data using `beautiful_soup`.

$ \ $


In [17]:
soup_game = BeautifulSoup(html_data, 'html5lib')

$ \ $

--------


Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Quarterly Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.




$ \ $


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [18]:
gme_revenue= pd.read_html(direction)[1]
gme_revenue.head()


GameStop Quarterly Revenue(Millions of US $)  \
0                                   2020-04-30   
1                                   2020-01-31   
2                                   2019-10-31   
3                                   2019-07-31   
4                                   2019-04-30   

  GameStop Quarterly Revenue(Millions of US $).1  
0                                         $1,021  
1                                         $2,194  
2                                         $1,439  
3                                         $1,286  
4                                         $1,548

In [19]:
gme_revenue=gme_revenue.rename(columns={"GameStop Quarterly Revenue(Millions of US $)":"Date","GameStop Quarterly Revenue(Millions of US $).1":"Revenue"})
gme_revenue.head()


Date Revenue
0  2020-04-30  $1,021
1  2020-01-31  $2,194
2  2019-10-31  $1,439
3  2019-07-31  $1,286
4  2019-04-30  $1,548

$ \ $ 

------

Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


$ \ $

In [20]:
gme_revenue.tail()

Date Revenue
57  2006-01-31  $1,667
58  2005-10-31    $534
59  2005-07-31    $416
60  2005-04-30    $475
61  2005-01-31    $709

In [21]:
gme_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")
gme_revenue.dropna(inplace=True)
gme_revenue = gme_revenue[tesla_revenue['Revenue'] != ""]
gme_revenue.reset_index(inplace=True)
gme_revenue.head()


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  """Entry point for launching an IPython kernel.


index        Date Revenue
0      0  2020-04-30   16934
1      1  2020-01-31   18756
2      2  2019-10-31   17719
3      3  2019-07-31   13757
4      4  2019-04-30   11958

$ \ $

-------

## *Question 5: Plot Tesla Stock Graph.*

Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


$ \ $



In [22]:
make_graph(tesla_data, tesla_revenue, 'Tesla')

$ \ $

-------------------

## *Question 6: Plot GameStop Stock Graph.*

Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


$ \ $




In [23]:
make_graph(gme_data, gme_revenue, 'GameStop')

<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork23455606-2022-01-01">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description          |
| ----------------- | ------- | ------------- | --------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part   |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab         |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
